src/main.rs

```diff
@@ -237,315 +241,277 @@ impl Format {
             },
             Format::Record(fields) => match fields.split_first() {
-                None => Lookahead::from(&next, len, Format::Empty),
-                Some(((_, a), fields)) => Lookahead::from(
-                    a,
-                    len,
-                    Format::Cat(Box::new(Format::Record(fields.to_vec())), Box::new(next)),
-                ),
+                None => self.add_next(index, depth, next),
+                Some(((_, a), fields)) => self.add(index, depth, a, &Next::Record(fields, next)),
             },
-            Format::Repeat(a) => Lookahead::from(
-                &Format::Alt(
-                    Box::new(Format::Empty),
-                    Box::new(Format::Cat(a.clone(), Box::new(Format::Repeat(a.clone())))),
-                ),
-                len,
-                next,
-            ),
+            Format::Repeat(a) => {
+                self.add_next(index, depth, next)?;
+                self.add(index, depth, a, &Next::Repeat(a, next))?;
+                Ok(())
+            }
             Format::RepeatCount(_expr, _a) => {
-                Some(Lookahead::empty()) // FIXME ?
+                self.accept(index) // FIXME
             }
             Format::Slice(_expr, _a) => {
-                Some(Lookahead::empty()) // FIXME ?
+                self.accept(index) // FIXME
             }
-            Format::WithRelativeOffset(_, _) => {
-                Some(Lookahead::empty()) // FIXME ?
+            Format::WithRelativeOffset(_expr, _a) => {
+                self.accept(index) // FIXME
             }
-            Format::Map(_f, a) => Lookahead::from(a, len, next),
+            Format::Map(_f, a) => self.add(index, depth, a, next),
             Format::If(_expr, a, b) => {
-                Lookahead::from(&Format::Alt(a.clone(), b.clone()), len, next)
+                self.add(index, depth, a, next)?;
+                self.add(index, depth, b, next)?;
+                Ok(())
             }
         }
     }
+
+    fn matches(&self, input: &[u8]) -> Option<usize> {
+        match input.split_first() {
+            None => self.accept,
+            Some((b, input)) => {
+                for (bs, s) in &self.branches {
+                    if bs.contains(*b) {
+                        return s.matches(input);
+                    }
+                }
+                self.accept
+            }
+        }
+    }
+
+    fn build0(depth: usize, branches: &[&Format], next: &Next) -> Option<Switch> {
+        let mut switch = Switch::reject();
+        for (i, branch) in branches.iter().enumerate() {
+            switch.add(i, depth, branch, next).ok()?;
+        }
+        Some(switch)
+    }
+
+    fn build(branches: &[&Format], next: &Next) -> Option<Switch> {
+        const MAX_DEPTH: usize = 16;
+        for depth in 0..MAX_DEPTH {
+            if let Some(switch) = Switch::build0(depth, branches, next) {
+                return Some(switch);
+            }
+        }
+        None
+    }
```

Comment:
I imagine this is trying `0..MAX_DEPTH` bytes of lookahead, trying a longer lookahead on failure?

Reply:
Yes, we want to construct the shallowest possible tree as deeper trees can take a very long time to construct on some pathological input. There may still be inputs that exhibit awful time complexity, unfortunately.